cpp/src/io/csv/csv_gpu.cu

```diff
@@ -570,6 +571,329 @@ __global__ void convertCsvToGdf(const char *raw_csv,
+          // Quoted string on newrow, or quoted string ending in terminator
+          ctx = make_short_row_context(ROW_CTX_QUOTE, ROW_CTX_NONE, ROW_CTX_QUOTE, 1, 0, 1);
+        } else {
+          // Start of a new row unless within a quote
+          ctx = make_short_row_context(ROW_CTX_NONE, ROW_CTX_QUOTE, ROW_CTX_NONE, 1, 0, 1);
+        }
+      } else if (c == quotechar) {
+        if (c_prev == delimiter || c_prev == quotechar) {
+          // Quoted string after delimiter, quoted string ending in delimiter, or double-quote
+          ctx = make_short_row_context(ROW_CTX_QUOTE, ROW_CTX_NONE);
+        } else {
+          // Closing or ignored quote
+          ctx = make_short_row_context(ROW_CTX_NONE, ROW_CTX_NONE);
+        }
+      } else if (c == delimiter_esc && c_prev != escapechar) {
+        // Escapable delimiter ends quoted string
+        ctx = make_short_row_context(ROW_CTX_NONE, ROW_CTX_NONE);
+      } else {
+        // Neutral character
+        ctx = make_short_row_context(ROW_CTX_NONE, ROW_CTX_QUOTE);
+      }
+    } else {
+      const char *data_end = start + data_size - start_offset;
+      if (cur >= data_end) {
+        // Add a newline at data end (need the extra row offset to infer length of previous row)
+        uint32_t eof_row = (cur == data_end && cur <= end) ? 1 : 0;
+        ctx =
+          make_short_row_context(ROW_CTX_EOF, ROW_CTX_EOF, ROW_CTX_EOF, eof_row, eof_row, eof_row);
+      } else {
+        // Pass-through context (beyond chunk_size but before data_size)
+        ctx = make_short_row_context(ROW_CTX_NONE, ROW_CTX_QUOTE, ROW_CTX_COMMENT);
+      }
+    }
+    // Merge with current context, keeping track of where new rows occur
+    merge_short_row_context(ctx_map, ctx, pos);
+  }
+
+  // Eliminate rows that start before byte_range_start
+  if (start_offset + block_pos < byte_range_start) {
+    uint32_t dist_minus1 = min(byte_range_start - (start_offset + block_pos) - 1, UINT64_C(31));
+    uint32_t mask        = 0xfffffffe << dist_minus1;
+    ctx_map.x &= mask;
+    ctx_map.y &= mask;
+    ctx_map.z &= mask;
+  }
+
+  // Convert the long-form {rowmap,outctx}[inctx] version into packed version
+  // {rowcount,ouctx}[inctx]
+  ctxb = pack_row_contexts(make_row_context(__popc(ctx_map.x), (ctx_map.w >> 0) & 3),
+                           make_row_context(__popc(ctx_map.y), (ctx_map.w >> 2) & 3),
+                           make_row_context(__popc(ctx_map.z), (ctx_map.w >> 4) & 3));
+
+  // Merge the row contexts of the 32-character blocks into a single 16K-character block context
+  rowctx_merge_transform(ctxtree, ctxb, t);
+
+  if (offsets_out) {
+    if (t == 0) { ctxtree[0] = row_ctx[blockIdx.x]; }
+    __syncthreads();
+
+    // Walk back the transform tree with the known initial parser state
+    rowctx32_t ctx             = rowctx_inverse_merge_transform(ctxtree, t);
+    uint64_t row               = (ctxtree[0] >> 2) + (ctx >> 2);
+    uint32_t rows_out_of_range = 0;
+    uint32_t rowmap =
```

Comment:
This could be an inline function `select_short_row_context()` maybe?

Reply:
Something like
```cpp
uint32_t select_short_row_context_count(uint4 ctx_map, rowctx32_t ctx) {
  return ((ctx & 3) == ROW_CTX_NONE)
        ? ctx_map.x
        : ((ctx & 3) == ROW_CTX_QUOTE) ? ctx_map.y : ((ctx & 3) == ROW_CTX_COMMENT) ? ctx_map.z : 0;
}
```

Reply:
Done.